File: Data Structures & Algorithms/delete-node-in-a-bst/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} key
     * @return {TreeNode}
     */
    deleteNode(root, key) {
        function deleteNodeHelper(root, key) {
            if (root === null) return root

            if (key < root.val) {
                root.left = deleteNodeHelper(root.left, key)
            } else if (key > root.val) {
                root.right = deleteNodeHelper(root.right, key)
            } else {
                if (root.left === null) {
                    return root.right
                } else if (root.right === null) {
                    return root.left
                } else {
                    const minNode = findAndReturnMinFromBST(root.right)
                    root.val = minNode.val
                    root.right = deleteNodeHelper(root.right, minNode.val)
                }
            }

            return root
        }

        return deleteNodeHelper(root, key)
    }
}


function findAndReturnMinFromBST(root) {
    let curr = root

    while (curr && curr.left) {
        curr = curr.left
    }

    return curr
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.py
class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def mainDfs(node):
            if not subRoot:
                return True
            if not node:
                return False
            
            if node.val == subRoot.val:
                if matchDfs(node, subRoot):
                    return True
            
            return mainDfs(node.left) or mainDfs(node.right)

        def matchDfs(n1, n2):
            if not n1 and not n2:
                return True
            if not n1 or not n2 or n1.val != n2.val:
                return False

            return matchDfs(n1.left, n2.left) and matchDfs(n1.right, n2.right)

        return mainDfs(root)